### ISSUE TITLE
Provide proper error message when Cruise Control fails to generate `KafkaRebalance` proposal

### ISSUE THREAD

**Issue by scholzj** [strimzi/strimzi-kafka-operator#8444]
When Cruise Control fails to generate a proposal, the operator doesn't really record any useful errors and user has to go into the Cruise Control logs to find out what the issue was. For example when the rebalance fails because of rack aware goals with following error:

```
2023-04-27 15:51:16 WARN  GoalOptimizer:606 - Detected unfixable proposal optimization
com.linkedin.kafka.cruisecontrol.exception.OptimizationFailureException: [RackAwareGoal] Insufficient number of racks to distribute each replica (Current: 1, Needed: 3). Add at least 2 racks with brokers. Add at least 2 racks with brokers.
	at com.linkedin.kafka.cruisecontrol.analyzer.goals.AbstractGoal.optimize(AbstractGoal.java:130) ~[cruise-control-2.5.121.jar:?]
	at com.linkedin.kafka.cruisecontrol.analyzer.GoalOptimizer.optimizations(GoalOptimizer.java:467) ~[cruise-control-2.5.121.jar:?]
	at com.linkedin.kafka.cruisecontrol.analyzer.GoalOptimizer.optimizations(GoalOptimizer.java:403) ~[cruise-control-2.5.121.jar:?]
	at com.linkedin.kafka.cruisecontrol.analyzer.GoalOptimizer$ProposalCandidateComputer.run(GoalOptimizer.java:599) ~[cruise-control-2.5.121.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
```

The operator seems to only report this:

```
2023-04-27 15:20:38 INFO  KafkaRebalanceAssemblyOperator:405 - Reconciliation #26(kafkarebalance-watch) KafkaRebalance(myproject/my-rebalance): Rebalance action is performed and KafkaRebalance resource is currently in [New] state
2023-04-27 15:20:38 INFO  KafkaRebalanceAssemblyOperator:1254 - Reconciliation #26(kafkarebalance-watch) KafkaRebalance(myproject/my-rebalance): Requesting Cruise Control rebalance [dryrun=true]
2023-04-27 15:20:38 INFO  CrdOperator:133 - Reconciliation #26(kafkarebalance-watch) KafkaRebalance(myproject/my-rebalance): Status of KafkaRebalance my-rebalance in namespace myproject has been updated
2023-04-27 15:20:38 INFO  KafkaRebalanceAssemblyOperator:405 - Reconciliation #27(kafkarebalance-watch) KafkaRebalance(myproject/my-rebalance): Rebalance action is performed and KafkaRebalance resource is currently in [PendingProposal] state
2023-04-27 15:20:43 INFO  KafkaRebalanceAssemblyOperator:1254 - Reconciliation #27(kafkarebalance-watch) KafkaRebalance(myproject/my-rebalance): Requesting Cruise Control rebalance [dryrun=true]
2023-04-27 15:20:48 INFO  KafkaRebalanceAssemblyOperator:1254 - Reconciliation #27(kafkarebalance-watch) KafkaRebalance(myproject/my-rebalance): Requesting Cruise Control rebalance [dryrun=true]
2023-04-27 15:20:48 ERROR KafkaRebalanceAssemblyOperator:880 - Reconciliation #27(kafkarebalance-watch) KafkaRebalance(myproject/my-rebalance): Cruise Control getting rebalance proposal failed
2023-04-27 15:20:48 ERROR KafkaRebalanceAssemblyOperator:485 - Reconciliation #27(kafkarebalance-watch) KafkaRebalance(myproject/my-rebalance): Status updated to [NotReady] due to error: null
2023-04-27 15:20:48 INFO  CrdOperator:133 - Reconciliation #27(kafkarebalance-watch) KafkaRebalance(myproject/my-rebalance): Status of KafkaRebalance my-rebalance in namespace myproject has been updated
2023-04-27 15:20:48 INFO  KafkaRebalanceAssemblyOperator:405 - Reconciliation #28(kafkarebalance-watch) KafkaRebalance(myproject/my-rebalance): Rebalance action is performed and KafkaRebalance resource is currently in [NotReady] state
```

We should be able to provide a better error than `null`.

**Comment by ShubhamRwt** [comment]
Hi Jakub, I have resolved this problem and currently what I have in the logs is this
```java
2023-05-02 14:49:29 INFO  KafkaRebalanceAssemblyOperator:405 - Reconciliation #1(test-trigger) KafkaRebalance(cruise-control-namespace/my-rebalance): Rebalance action is performed and KafkaRebalance resource is currently in [PendingProposal] state
2023-05-02 14:49:34 INFO  KafkaRebalanceAssemblyOperator:1254 - Reconciliation #1(test-trigger) KafkaRebalance(cruise-control-namespace/my-rebalance): Requesting Cruise Control rebalance [dryrun=true]
2023-05-02 14:49:34 ERROR KafkaRebalanceAssemblyOperator:880 - Reconciliation #1(test-trigger) KafkaRebalance(cruise-control-namespace/my-rebalance): Cruise Control getting rebalance proposal failed
io.strimzi.operator.cluster.operator.resource.cruisecontrol.CruiseControlRestException: Rebalance returned unknown response: User Task ID: rebalance-no-goals-response JSON: {"version":1,"progress":[{"operation":"Rebalance","operationProgress":[{"description":"Operation enqueued, waiting to be executed.","completionPercentage":100.0,"step":"PENDING","time-in-ms":127},{"description":"The job requires a cluster model and it is waiting to get the cluster model lock.","completionPercentage":100.0,"step":"WAITING_FOR_CLUSTER_MODEL","time-in-ms":461}]}]}
	at io.strimzi.operator.cluster.operator.assembly.KafkaRebalanceAssemblyOperator.handleRebalanceResponse(KafkaRebalanceAssemblyOperator.java:1309) ~[classes/:?]
	at io.strimzi.operator.cluster.operator.assembly.KafkaRebalanceAssemblyOperator.lambda$requestRebalance$36(KafkaRebalanceAssemblyOperator.java:1276) ~[classes/:?]
	at io.vertx.core.impl.future.Mapping.onSuccess(Mapping.java:35) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.future.Composition$1.onSuccess(Composition.java:62) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.future.SucceededFuture.addListener(SucceededFuture.java:88) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.future.Composition.onSuccess(Composition.java:43) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:60) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.Promise.complete(Promise.java:66) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.strimzi.operator.cluster.operator.resource.cruisecontrol.CruiseControlApiImpl.lambda$internalRebalance$4(CruiseControlApiImpl.java:180) ~[classes/:?]
	at io.vertx.core.impl.future.FutureImpl$1.onSuccess(FutureImpl.java:91) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.future.FutureBase.lambda$emitSuccess$0(FutureBase.java:54) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:86) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.ContextBase.execute(ContextBase.java:225) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.future.FutureBase.emitSuccess(FutureBase.java:51) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.future.FutureImpl.tryComplete(FutureImpl.java:211) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.future.PromiseImpl.tryComplete(PromiseImpl.java:23) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.http.impl.HttpEventHandler.handleEnd(HttpEventHandler.java:79) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.http.impl.HttpClientResponseImpl.handleEnd(HttpClientResponseImpl.java:250) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.lambda$new$0(Http1xClientConnection.java:435) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.streams.impl.InboundBuffer.handleEvent(InboundBuffer.java:239) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.streams.impl.InboundBuffer.write(InboundBuffer.java:129) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.http.impl.Http1xClientConnection$StreamImpl.handleEnd(Http1xClientConnection.java:682) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.EventLoopContext.execute(EventLoopContext.java:76) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.impl.ContextBase.execute(ContextBase.java:232) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.http.impl.Http1xClientConnection.handleResponseEnd(Http1xClientConnection.java:916) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.http.impl.Http1xClientConnection.handleHttpMessage(Http1xClientConnection.java:788) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.http.impl.Http1xClientConnection.handleMessage(Http1xClientConnection.java:752) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.net.impl.ConnectionBase.read(ConnectionBase.java:157) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.vertx.core.net.impl.VertxHandler.channelRead(VertxHandler.java:153) ~[vertx-core-4.3.8.jar:4.3.8]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:436) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:346) ~[netty-codec-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:318) ~[netty-codec-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:251) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:442) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1373) ~[netty-handler-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1236) ~[netty-handler-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1285) ~[netty-handler-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:529) ~[netty-codec-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:468) ~[netty-codec-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:290) ~[netty-codec-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:444) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:412) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1410) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:440) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:420) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:919) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:166) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:788) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:724) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:650) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:562) ~[netty-transport-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.util.concurrent.SingleThreadEventExecutor$4.run(SingleThreadEventExecutor.java:997) ~[netty-common-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.util.internal.ThreadExecutorMap$2.run(ThreadExecutorMap.java:74) ~[netty-common-4.1.87.Final.jar:4.1.87.Final]
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30) ~[netty-common-4.1.87.Final.jar:4.1.87.Final]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
2023-05-02 14:49:34 ERROR KafkaRebalanceAssemblyOperator:485 - Reconciliation #1(test-trigger) KafkaRebalance(cruise-control-namespace/my-rebalance): Status updated to [NotReady] due to error: Rebalance returned unknown response: User Task ID: rebalance-no-goals-response JSON: {"version":1,"progress":[{"operation":"Rebalance","operationProgress":[{"description":"Operation enqueued, waiting to be executed.","completionPercentage":100.0,"step":"PENDING","time-in-ms":127},{"description":"The job requires a cluster model and it is waiting to get the cluster model lock.","completionPercentage":100.0,"step":"WAITING_FOR_CLUSTER_MODEL","time-in-ms":461}]}]}
```

**Comment by scholzj** [comment]
That does not seem to be exactly user friendly either TBH. Are those logs for the same error I got? Or is it for some other issue?

**Comment by ShubhamRwt** [comment]
I was testing with a different scenario. My doubt was that, Do we need the error message/trace with this log also 
```
2023-04-27 15:20:48 ERROR KafkaRebalanceAssemblyOperator:880 - Reconciliation #27(kafkarebalance-watch) KafkaRebalance(myproject/my-rebalance): Cruise Control getting rebalance proposal failed : Status updated to [NotReady] due to error: Rebalance returned unknown response: User Task ID: rebalance-no-goals-response JSON: {"version":1,"progress":[{"operation":"Rebalance","operationProgress":[{"description":"Operation enqueued, waiting to be executed.","completionPercentage":100.0,"step":"PENDING","time-in-ms":127},{"description":"The job requires a cluster model and it is waiting to get the cluster model lock.","completionPercentage":100.0,"step":"WAITING_FOR_CLUSTER_MODEL","time-in-ms":461}]}]}

2023-05-02 14:49:34 ERROR KafkaRebalanceAssemblyOperator:485 - Reconciliation #1(test-trigger) KafkaRebalance(cruise-control-namespace/my-rebalance): Status updated to [NotReady] due to error: Rebalance returned unknown response: User Task ID: rebalance-no-goals-response JSON: {"version":1,"progress":[{"operation":"Rebalance","operationProgress":[{"description":"Operation enqueued, waiting to be executed.","completionPercentage":100.0,"step":"PENDING","time-in-ms":127},{"description":"The job requires a cluster model and it is waiting to get the cluster model lock.","completionPercentage":100.0,"step":"WAITING_FOR_CLUSTER_MODEL","time-in-ms":461}]}]}
```

**Comment by ShubhamRwt** [comment]
I mean both these line are reporting same error message so do we need it in both the logs?

**Comment by scholzj** [comment]
I guess you don't need two errors. But at the same time, it is better to get in some cases 2 errors than in some cases no errors. But the main issue is for me with the JSON -> that is fine in some DEBUG level messages. But for ERROR logs, we should try to have a nice readable message, so we should probably parse it. Also, I'm not sure what did you do to get this -> the JOSN seems very confusing as to what happened ... I would read it as everything is in progress.

**Comment by ShubhamRwt** [comment]
I didn't changed the code much, just changed the `e.getCause()` to just `e` and everything started working. Regarding the json part it was like that. This is the error which is thrown by `CruiseControlApilImpl`. 
```java
                                result.fail(new CruiseControlRestException(
                                        "Error for request: " + host + ":" + port + path +
                                                ". 202 Status code did not contain progress key. Server returned: " +
                                                ccResponse.getJson().toString()));
```

**Comment by scholzj** [comment]
Well, the question is what does changing from `e.getCause()` to `e` mean in other cases. As I guess that in some situations, `e.getCause` is the right thing and that is why it is used?

**Comment by ShubhamRwt** [comment]
`e.getCause()` is coming out as null in this case since we are generating our own exception classes and  we have not providing any cause there. So therefore we have to use `e` directly here.

**Comment by scholzj** [comment]
Yeah -> but the question is why was `e.getCause()` used there in the first place. Could it be that there are some other use-cases (errors) where `e.getCause()` is correct and you break it by using just `e`?

**Comment by ShubhamRwt** [comment]
Fixed in #8588
